pints/_mcmc/_smc.py

```diff
@@ -0,0 +1,266 @@
+        return self._tt_evaluate(x)
+
+    def run(self):
+        from tempfile import mkdtemp
+
+        # TODO: allow changes before run
+        # number of chains
+        n_chains = 100
+
+        # number of steps
+        n_steps = 20
+
+        # tuning interval
+        tune_interval = 5
+
+        # number of jobs
+        n_jobs = 10
+
+        # starting stage
+        stage = 0
+
+        # initial spread
+        sigma_x0 = 0.2  # float or array_like of floats
+
+        # output folder
+        homepath = mkdtemp(prefix='PINTS_SMC')
+
+        # random seed
+        seed = 21
+
+        # Starting point
+        x0 = self._x0
+
+        # Start SMC setup
+        pm_model = pm.Model()
+        with pm_model:
+            # setup prior as PyMC3
+            # pm_prior = self._get_prior_as_pm(self._log_likelihood._prior)
+            # pm_sigma = self._get_sigma_as_pm(self._log_likelihood._prior)
+
+            # setup model as PyMC3
+            # Assuming we are doing pints.BayesianLogLikelihood
+            # mu = self._tt_evaluate(pm_prior)
+            # cov = np.eye(1) * pm_sigma**2
+
+            # setup likelihood as PyMC3
+            # Assuming Guassian
+            # pm_likelihood = pm.MvNormal('likelihood', mu=mu, cov=cov,
+            #                            observed=self._values)
+
+            # Guess x0
+            start = [
+                {'prior': np.random.normal(x0[:-1], x0[:-1] * sigma_x0),
+                 'sigma': np.random.normal(x0[-1], x0[-1] * sigma_x0)}
+                for _ in range(n_chains)]
+
+            # run!
+            pm_chain = pm.step_methods.smc.sample_smc(
+                n_steps=n_steps,
+                n_chains=n_chains,
+                progressbar=self._verbose,
+                homepath=homepath,
+                n_jobs=n_jobs,
+                stage=stage,
+                start=start,
+                tune_interval=tune_interval,
+                random_seed=seed
+            )
+
+        # Get marginal likelihood
+        self._marginal_likelihood = pm_model.marginal_likelihood
+
+        # Convert it to numpy
+        from pymc3.plots.utils import get_default_varnames
+        varname = get_default_varnames(pm_chain.varnames, False)
+        shape = pm_chain.get_values(
+            varname[0], combine=True, squeeze=True).shape
+        chain = np.zeros((shape[0], shape[1] + 1))
+        chain[:, :-1] = pm_chain.get_values(
+            varname[0], combine=True, squeeze=True)
+        chain[:, -1] = pm_chain.get_values(
+            varname[1], combine=True, squeeze=True)
+
+        # Return generated chain
+        return chain
+
+    def marginal_likelihood(self):
+        """
+        Return the computed marginal likelihood through the SMC routine.
+
+        Return None if SMC.run() not yet run.
+        """
+        return self._marginal_likelihood
+
+
+@as_op(itypes=[tt.dvector], otypes=[tt.dmatrix])
+def _tt_evaluate_single_series(x):
+    sol = evaluate(x)
```

Comment:
What's going on here? Should the evaluation be inside the try/catch? What kind of exception are we hoping to catch?

Reply:
Since we had the model returning `inf` if we cannot solve it with the given parameters, then I put this `try` to temporary get away from that. But it should be something else really... (We should either raise the exception as `ValueError` and ask the user (myself) to change the return to something else when a parameter is not solvable -- but what should the something else be(?). Or we explicitly say return `inf` if the user-defined model becomes not-solvable for the given parameters, and we deal with it better here.)

Reply:
Having the `evaluate(x)` inside try/catch does not help I think, because the `evaluate(.)` does not return any error, but yes, we can put it inside (it probably looks better).

Reply:
I still don't understand what the try is supposed to catch